File: Programmers/python/프로세스.py
from collections import deque

def update_location(location, dq):
    if location == 0:
        return len(dq) - 1
    return location - 1

def solution(priorities, location):
    dq = deque(priorities)
    order = 0

    while dq:
        # 가장 높은 우선순위 찾기
        highest = max(dq)
        target = dq.popleft()
        
        # 더 큰 프로세스가 있는 경우 
        if target < highest:
            dq.append(target)
        else:
            # append를 안했다는 것은 프로세스가 실행되었다는 것
            order += 1
            if location == 0:
                return order
            
        location = update_location(location, dq)

File: Programmers/java/프로세스.java
import java.util.*;

class Solution {
    public int solution(int[] priorities, int location) {
        int answer = 0;
        
        Queue<int[]> queue = new LinkedList<>();
        
        // 큐에 인덱스와 우선순위를 함께 저장
        for (int i = 0; i < priorities.length; i++) {
            queue.add(new int[] {i, priorities[i]});
        }
        
        while (!queue.isEmpty()) {
            int[] cur = queue.poll();
            int max = findMax(queue); // 큐에서 가장 큰 우선순위 확인
            
            // 현재보다 우선순위가 높다면 다시 큐에 추가
            if (cur[1] < max) {
                queue.add(cur);
            } else {
                answer++;
                
                // 찾던 프로세스라면 정답 반환
                if (cur[0] == location) {
                    return answer;
                }
            }
        }
        
        return answer;
    }
    
    // 큐에서 최댓값을 찾는 함수
    public int findMax(Queue<int[]> queue) {
        int max = Integer.MIN_VALUE;
        
        for (int[] value : queue) {
            if (value[1] > max) {
                max = value[1];
            }
        }
        
        return max;
    }
}
